javalib/src/main/scala/java/math/BigDecimal.scala

```diff
@@ -0,0 +1,1761 @@
+              exponent -= k
+              mantissa0.shiftLeft(k)
+            } else {
+              mantissa0
+            }
+          }
+          // Computing (mantissa * 2^k) / 10^s
+          val qr: QuotAndRem[BigInteger] = m.divideAndRemainder(powerOfTen)
+          // To check if the fractional part >= 0.5
+          val compRem = qr.rem.shiftLeftOneBit().compareTo(powerOfTen)
+          // To add two rounded bits at end of mantissa
+          exponent -= 2
+          qr.quot.shiftLeft(2).add(BigInteger.valueOf((compRem * (compRem + 3)) / 2 + 1))
+        }
+      }
+
+      val lowestSetBit = mantissa.getLowestSetBit
+      val discardedSize = mantissa.bitLength() - 54
+      var bits: Long = 0L // IEEE-754 Standard
+      var tempBits: Long = 0L // for temporal calculations
+      if (discardedSize > 0) { // (#bits > 54)
+        bits = mantissa.shiftRight(discardedSize).longValue()
+        tempBits = bits
+        if (((bits & 1) == 1 && lowestSetBit < discardedSize) || (bits & 3) == 3)
+          bits += 2
+      } else { // (#bits <= 54)
+        bits = mantissa.longValue() << -discardedSize
+        tempBits = bits
+        if ((bits & 3) == 3)
+          bits += 2
+      }
+      // Testing bit 54 to check if the carry creates a new binary digit
+      if ((bits & 0x40000000000000L) == 0) {
+        // To drop the last bit of mantissa (first discarded)
+        bits >>= 1
+        exponent += discardedSize
+      } else {
+        // #bits = 54
+        bits >>= 2
+        exponent += (discardedSize + 1)
+      }
+      // To test if the 53-bits number fits in 'double'
+      if (exponent > 2046) {
+        // (exponent - bias > 1023)
+        sign * Double.PositiveInfinity
+      } else if (exponent < -53) {
+        sign * 0.0d
+      } else {
+        if (exponent <= 0) {
+          bits = tempBits >> 1
+          tempBits = bits & (-1L >>> (63 + exponent))
+          bits >>= (-exponent)
+          // To test if after discard bits, a new carry is generated
+          if (((bits & 3) == 3) ||
+              (((bits & 1) == 1) && (tempBits != 0) && (lowestSetBit < discardedSize))) {
+            bits += 1
+          }
+          exponent = 0
+          bits >>= 1
+        }
+
+        // Construct the 64 double bits: [sign(1), exponent(11), mantissa(52)]
+        val resultBits =
+          (sign & 0x8000000000000000L) |
+          (exponent.toLong << 52)      |
+          (bits & 0xFFFFFFFFFFFFFL)
+        java.lang.Double.longBitsToDouble(resultBits)
+      }
+    }
+  }
+
+  def ulp(): BigDecimal = valueOf(1, _scale)
+
+  private def decimalDigitsInLong(value: Long): Int = {
+    if (value == Long.MinValue) {
+      19 // special case required because abs(MIN_VALUE) == MIN_VALUE
+    } else {
+      val index = Arrays.binarySearch(LongTenPows, Math.abs(value))
+      if (index < 0) -index - 1
+      else index + 1
+    }
+  }
+
+  @inline
+  private def divideAndRemainderImpl(divisor: BigDecimal): Array[BigDecimal] = {
```

Comment:
The result type should be `QuoteAndRem` here. (example of what an implicit conversion was not a good idea)